packages/box/src/dfuse-receiver.ts

```diff
@@ -0,0 +1,242 @@
+                                hexData: a.hexData,
+                            })),
+                        });
+                    }
+                    this.storage.pushJsonBlock(
+                        JSON.stringify(block),
+                        prev.irreversibleBlockNum
+                    );
+                }
+                this.unpushedTransactions = [];
+            }
+        }
+        if (trx.trace !== null) this.unpushedTransactions.push(trx);
+    }
+
+    onMessage(message: GraphqlStreamMessage<any>, stream: Stream): void {
+        if (message.type === "data") {
+            const trx: JsonTrx = message.data.searchTransactionsForward;
+            const prev =
+                this.jsonTransactions.length > 0
+                    ? this.jsonTransactions[this.jsonTransactions.length - 1]
+                    : null;
+            logger.info(
+                `${trx.undo ? "undo block" : "recv block"} ${trx.block.num} ${
+                    trx.trace
+                        ? "trx " + trx.trace.id
+                        : "no matching transactions"
+                }`
+            );
+            if (trx.trace || (prev && prev.trace)) {
+                this.jsonTransactions.push(trx);
+                this.pushTrx(trx);
+                if (
+                    this.jsonTransactions.length - this.numSaved > 10 ||
+                    !trx.trace
+                ) {
+                    logger.info(
+                        `save ${dfuseConfig.jsonTrxFile}: ${this.jsonTransactions.length} transactions and undo entries`
+                    );
+                    fs.writeFileSync(
+                        dfuseConfig.jsonTrxFile + ".tmp",
+                        JSON.stringify(this.jsonTransactions)
+                    );
+                    fs.renameSync(
+                        dfuseConfig.jsonTrxFile + ".tmp",
+                        dfuseConfig.jsonTrxFile
+                    );
+                    this.storage.saveState();
+                    this.numSaved = this.jsonTransactions.length;
+                }
+            }
+            stream.mark({
+                cursor: trx.cursor,
+            });
+        }
+        if (message.type === "error")
+            logger.error(JSON.stringify(message, null, 4));
+    }
+
+    async start() {
+        try {
+            this.jsonTransactions = JSON.parse(
+                fs.readFileSync(dfuseConfig.jsonTrxFile, "utf8")
+            );
+            this.numSaved = this.jsonTransactions.length;
+        } catch (e) {}
+
+        if (this.jsonTransactions.length)
+            this.variables.cursor = this.jsonTransactions[
+                this.jsonTransactions.length - 1
+            ].cursor;
+
+        const begin = performance.now();
+        logger.info("pushing existing blocks...");
+        for (let trx of this.jsonTransactions) this.pushTrx(trx);
+        logger.info(`${performance.now() - begin} ms`);
+        this.storage.saveState();
```

Comment:
so every time the server restarts we read all the json transactions and overwrite the current storage state? if not, could you please explain me how it works?

Reply:
Correct. We don't have a stable state format yet, so it plays it safe for now.